### PR TITLE
Add a standard format for archiving build results.

### PR DESCRIPTION
**Pull request by nuclearsandwich**
The first commit adds documentation describing a proposed build-results-archive file.

Further commits will add to this python module API methods for building the described archive.

I'm opening this as a draft but I am looking for feedback on the proposed format.

### CODE REVIEW

File: docs/build-results-archive.md
Comment (mjeronimo) [inline docs/build-results-archive.md:43]: Should we also state that the files should be read-only, just to reinforce the idea that they are immutable?

File: docs/build-results-archive.md
Comment (mjeronimo) [inline docs/build-results-archive.md]: Once we have a "colcon sarif" command (whatever we decide the specific verb name should be), this could become "colcon-sarif-cmd", and would then be unform with "colcon-build-cmd", and "colcon-test-cmd".

**Review by mjeronimo — APPROVED** [review]

File: docs/build-results-archive.md
Comment (Ronoman) [inline docs/build-results-archive.md:1]: Should we specify where this results archive is written to?

File: docs/build-results-archive.md
Comment (Ronoman) [inline docs/build-results-archive.md]: Nitpick: `should`, or `shall`?

Comment (Ronoman) [reply docs/build-results-archive.md]: Won't repeat the review comment, but there are more instances of `should` in the doc to check. Also either all {should, shall, must} instances should be wrapped with `_`, or none of them (though happy to be overridden on this one).

File: docs/build-results-archive.md
Comment (Ronoman) [inline docs/build-results-archive.md]: Is this something we want to have? Merging Sarif files is non-trivial, but we could implement it.

Pros:
- Single file to open for the Dashboard to process and produce a view for
- Single place for developers to look for output if they want a manual view

Cons:
- A bit tricky to implement
    - What to do with multiple tools?
    - Need to merge Artifact lists and update Result indices accordingly
    - Need to ensure SARIF metadata is set properly (high level `properties` field)

It may be as simple as appending multiple objects to the `runs` list, but currently `process_sarif` assumes a single run per SARIF file.

File: docs/build-results-archive.md
Comment (nuclearsandwich) [reply docs/build-results-archive.md:43]: I think that the archive itself should be entirely read-only and immutable. It's worth noting that in the text but whether the files themselves are "read-only" or not will be implementation dependent. (Ours will probably try to make them as such)

File: docs/build-results-archive.md
Comment (mayman99) [reply docs/build-results-archive.md]: +1 for not merging the files

File: docs/build-results-archive.md
Comment (nuclearsandwich) [reply docs/build-results-archive.md]: It could although I'm not entirely convinced that it should be a colcon command or that even if it is a colcon command that it needs to be and therefore semantically I would argue against putting colcon in the filename. In fact I'd probably sooner argue that we should strip it from the build and test command filenames and call them `build-cmd` and `test-cmd` except that there really isn't a substitute for colcon in those cases and I think it's worth separating that definition from the build and test commands for any one package.

If we want to bikeshed a little perhaps `workspace-{build,test}-cmd`?

File: docs/build-results-archive.md
Comment (nuclearsandwich) [reply docs/build-results-archive.md]: Okay it sounds like from out-loud discussion that keeping the same file structure (one output per tool per package) in the processed sarif is the way to go. This means that rather than have one processed.sarif file we'll have a directory mirroring the build directory but with de-duplicated sarif files.

Comment (nuclearsandwich) [reply docs/build-results-archive.md]: https://github.com/space-ros/process_sarif/pull/6/commits/2d3b4fa30c36f6680f170652e8e1f198868ed6aa updates the doc to specify a processed directory rather than a single file.

File: docs/build-results-archive.md
Comment (nuclearsandwich) [reply docs/build-results-archive.md]: Good catch on the should vs shall/must, I fixed that in https://github.com/space-ros/process_sarif/pull/6/commits/f3144142872728a02d0d7281cd6ae42dcaf9a5d0 along with emphasizing RFC words consistently.

File: docs/build-results-archive.md
Comment (nuclearsandwich) [reply docs/build-results-archive.md:43]: Noted in https://github.com/space-ros/process_sarif/pull/6/commits/b4440873d32415b983db5dbf23889cdfe06c9108 along with a suggestion that applications using the data use their own formats rather than alter these files.

**Comment by Ronoman** [comment]
Added vcs exact metadata, and swapped out the globbing list comprehension for a sarif_helpers function which handles malformed SARIF better.

File: process_sarif/build_results_archive.py
Comment (nuclearsandwich) [inline process_sarif/build_results_archive.py]: At the very least, it's probably worth making this a `check_call()` so that the entire process fails if the vcs-export does. I had been thinking to use vcs as a [runnable module](https://peps.python.org/pep-0338/) rather than shelling out directly but I'm not sure that it's ultimately that important

File: docs/build-results-archive.md
Comment (mayman99) [reply docs/build-results-archive.md:1]: It is specified in the next section :+1:

**Review by mayman99 — APPROVED** [review]

File: docs/build-results-archive.md
Comment (nuclearsandwich) [reply docs/build-results-archive.md:1]: @Ronoman as mentioned these lines deal with the path and filename conventions: https://github.com/space-ros/process_sarif/pull/6/files#diff-c55023a2efb41f70f1da8ee83deee86689f8bc5e843e745e305ccd8c750c5defR11-R15

**Comment by nuclearsandwich** [comment]
@Ronoman while I'm waiting for your review would you prefer me merge this PR after that review or would you prefer to be the one who merges PRs on this project for the time being?

The ROS 2 core team's convention is that team members merge their own PRs (with a "you break it you bought it" motivation) but I'm happy to do otherwise if you prefer

File: docs/build-results-archive.md
Comment (Ronoman) [inline docs/build-results-archive.md]: Is this left behind from a previous iteration? My understanding is we decided for multiple `.sarif` files (corroborated below in processed/**/*.sarif), but this might be for something else that I'm forgetting about.

File: process_sarif/build_results_archive.py
Comment (Ronoman) [inline process_sarif/build_results_archive.py]: Does this work? My intent with this function is that it doesn't modify its inputs and instead returns `List[SarifFile]` that have duplicated removed, but its possible that [the function](https://github.com/space-ros/process_sarif/blob/main/process_sarif/sarif_helpers.py#L88) does in fact modify the argument in-place.

**Review by Ronoman — CHANGES_REQUESTED** [review]
Two very small changes to address before merging, but this looks great.

**Comment by Ronoman** [comment]
Since these are your changes, I'll let you merge the PR. In the future I don't have a preference, but the ROS2 core team convention seems reasonable to me.

File: process_sarif/build_results_archive.py
Comment (nuclearsandwich) [reply process_sarif/build_results_archive.py]: Good catch! This is left over from one of my earlier drafts fixing up the remove_duplicates function which did modify in place. ebdca25 saves the processed files and writes _them_ to the processed subdir instead of the original SARIF files.

File: docs/build-results-archive.md
Comment (nuclearsandwich) [reply docs/build-results-archive.md]: corrected in e7eded4

**Review by Ronoman — APPROVED** [review]

**Pull request merged**
